Modules/_ctypes/_ctypes.c

```diff
@@ -956,32 +956,47 @@ CDataType_in_dll_impl(PyObject *type, PyTypeObject *cls, PyObject *dll,
         return NULL;
     }
 
+#undef USE_DLERROR
 #ifdef MS_WIN32
     Py_BEGIN_ALLOW_THREADS
     address = (void *)GetProcAddress(handle, name);
     Py_END_ALLOW_THREADS
-    if (!address) {
-        PyErr_Format(PyExc_ValueError,
-                     "symbol '%s' not found",
-                     name);
-        return NULL;
-    }
 #else
+    #ifdef __CYGWIN__
+        // dlerror() isn't very helpful on cygwin
+    #else
+        #define USE_DLERROR
+        /* dlerror() always returns the latest error.
+         *
+         * Clear the previous value before calling dlsym(),
+         * to ensure we can tell if our call resulted in an error.
+         */
+        (void)dlerror();
+    #endif
     address = (void *)dlsym(handle, name);
-    if (!address) {
-#ifdef __CYGWIN__
-/* dlerror() isn't very helpful on cygwin */
-        PyErr_Format(PyExc_ValueError,
-                     "symbol '%s' not found",
-                     name);
-#else
-        PyErr_SetString(PyExc_ValueError, dlerror());
 #endif
-        return NULL;
+
+    if (address) {
+        ctypes_state *st = get_module_state_by_def(Py_TYPE(type));
+        return PyCData_AtAddress(st, type, address);
+    }
+
+#ifdef USE_DLERROR
+    const char *dlerr = dlerror();
+    if (dlerr) {
+        PyObject *message = PyUnicode_DecodeLocale(dlerr, "surrogateescape");
+        if (message) {
+            PyErr_SetObject(PyExc_ValueError, message);
+            return NULL;
+        }
```

Comment:
This will ignore exceptions from `PyUnicode_DecodeLocale` (since the `PyErr_Format` below will overwrite the current exception). If that's the intent, there should be an explicit `PyErr_Clear`.

```suggestion
            PyErr_SetObject(PyExc_ValueError, message);
            return NULL;
        }
        // Ignore errors from PyUnicode_DecodeLocale,
        // fall back to the generic error below.
        PyErr_Clear();
```
I still think that this (with `PyErr_Clear`) would be best: the generic `symbol '%s' not found` is more helpful than “sorry, we couldn't produce your error message”. (And unlike Python, a `KeyboardInterrupt` can't appear at an arbitrary point in C code -- we'd really be *only* ignoring `PyUnicode_DecodeLocale` errors.)

---

But if the intent is to propagate any exception from `PyUnicode_DecodeLocale` (instead of the `ValueError("symbol not found")`), that would be fine too. That would need a `return NULL` after the *if* body, though.

(same 2× below)

Reply:
Yeah, but `symbol not found` might be very misleading to users, especially if the library actually has the symbol. Instead of clearing the error, could we just call `PyErr_WriteUnraisable` so the user at least sees *something*?

Reply:
I agree with @ZeroIntensity that falling back to `symbol not found` may be misleading in some cases, especially if the cause of the error is not the absence of the symbol (or an IFUNC returning NULL), but something else is going on.

So, I think we should propagate the exception, and maybe accompany it with the non-decoded string, if that makes any sense. Would `PyErr_WriteUnraisable` help in this case?

Reply:
IMO, `symbol not found` is correct: if `dlsym` gives us NULL, we *can* say that Python did not find the symbol. (We'd like to say *why*, but if we fail to get that reason, a generic error is fine. It better be fine -- it's what Windows/Cygwin users will see.)
In other words: if an IFUNC tells us that the symbol is missing (NULL), well, that means we can't find it :)

But if you disagree: the error from `PyUnicode_DecodeLocale` is raisable. If it's helpful to the user, then raise it (i.e. return `NULL`).

Reply:
Your point makes sense. From the scope of the callee (Python user), it's indistinguishable.

Let's not complicate things further.

@ZeroIntensity what do you think?

Reply:
I'm not the biggest fan of swallowing errors, but I'll defer to Petr. (That means, go ahead and commit his suggestion and I'll approve again.)

Reply:
Done and pushed!

I also tried indenting the `#ifdef USE_DLERROR` / `#undef` according to https://github.com/python/cpython/pull/126555#discussion_r1838310417